src/theory/uf/proof_equality_engine.h

```diff
@@ -0,0 +1,340 @@
+   * holds in the equality engine, this method returns false.
+   */
+  bool assertFact(Node lit, Node exp, ProofStepBuffer& psb);
+  /**
+   * Assert fact via generator pg. This method asserts lit with explanation exp
+   * to the equality engine of this class. It must be the case that pg can
+   * provide a proof for lit in terms of exp. More precisely, pg should be
+   * prepared in the remainder of the SAT context to respond to a call to a
+   * call to ProofGenerator::getProofFor(lit), and return a proof whose free
+   * assumptions are a subset of the conjuncts of exp.
+   *
+   * @param lit The literal to assert to the equality engine.
+   * @param exp The premises of the proof concluding lit. These are also
+   * the premises that are used when calling explain(lit).
+   * @param pg The proof generator that can provide a proof concluding lit
+   * from free asumptions in exp.
+   * @return true if this fact was processed by this method. If lit already
+   * holds in the equality engine, this method returns false.
+   */
+  bool assertFact(Node lit, Node exp, ProofGenerator* pg);
+  //-------------------------- assert conflicts
+  /**
+   * This method is called when the equality engine of this class is
+   * inconsistent (false has been proven) by a contradictory literal lit. This
+   * returns the trust node corresponding to the current conflict.
+   *
+   * @param lit The conflicting literal, which must rewrite to false.
+   * @return The trust node capturing the fact that this class can provide a
+   * proof for this conflict.
+   */
+  TrustNode assertConflict(Node lit);
+  /**
+   * Get proven conflict from contradictory facts. This method is called when
+   * the proof rule with premises exp and arguments args implies a contradiction
+   * by proof rule id.
+   *
+   * This method returns the TrustNode containing the corresponding conflict
+   * resulting from adding this step, and ensures that a proof has been stored
+   * internally so that this class may respond to a call to
+   * ProofGenerator::getProof(...).
+   */
+  TrustNode assertConflict(PfRule id,
+                           const std::vector<Node>& exp,
+                           const std::vector<Node>& args);
+  /** Multi-step version */
+  TrustNode assertConflict(const std::vector<Node>& exp, ProofStepBuffer& psb);
+  //-------------------------- assert lemma
+  /**
+   * Called when we have concluded conc, typically via theory specific
+   * reasoning. The purpose of this method is to construct a TrustNode of
+   * kind TrustNodeKind::LEMMA or TrustNodeKind::CONFLICT corresponding to the
+   * lemma or conflict to be sent on the output channel of the Theory.
+   *
+   * The user provides the explanation of conc in two parts:
+   * (1) exp \ noExplain, which are literals that hold in the equality engine of
+   * this class,
+   * (2) noExplain, which do not necessarily hold in the equality engine of this
+   * class.
+   * Notice that noExplain is a subset of exp.
```

Comment:
This is contradictory with exp being literals that hold in the equality engine and noExplain being literals that do not.

Reply:
The first says `exp \ noExplain`, I'll parenthesize.